kroxylicious-filters/kroxylicious-user-namespace/src/main/templates/UserNamespaceFilter.ftl

```diff
@@ -0,0 +1,354 @@
+import static org.apache.kafka.common.protocol.ApiKeys.DESCRIBE_GROUPS;
+import static org.apache.kafka.common.protocol.ApiKeys.FIND_COORDINATOR;
+import static org.apache.kafka.common.protocol.ApiKeys.JOIN_GROUP;
+import static org.apache.kafka.common.protocol.ApiKeys.OFFSET_COMMIT;
+
+
+/**
+* Decodes Kafka Readable into an ApiMessage
+* <p>Note: this class is automatically generated from a template</p>
+*/
+public class UserNamespaceFilter implements RequestFilter, ResponseFilter {
+
+    private final UserNamespace.Config config;
+
+    private static final Set<ApiKeys> responseKeys = Set.of(FIND_COORDINATOR, OFFSET_COMMIT, CONSUMER_GROUP_DESCRIBE, DESCRIBE_GROUPS, JOIN_GROUP);
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserNamespaceFilter.class);
+
+    <#list messageSpecs as messageSpec>
+        <#if messageSpec.type?lower_case == 'request' && messageSpec.entityFields.hasAtLeastOneEntityField>
+            <#assign specName>
+            <#assign words=messageSpec.name?split("(?=[A-Z])", "r")><#list words as word>${word?c_upper_case}<#sep>_</#list>
+            </#assign>
+    private static final Set<Short> ${specName?trim}_VERSIONS = Set.of(<#list messageSpec.entityFields.orderedVersions as version>(short) ${version}<#sep>, </#list>);
+        </#if>
+    </#list>
+
+    UserNamespaceFilter(UserNamespace.Config config) {
+        this.config = config;
+    }
+
+    @Override
+    public boolean shouldHandleRequest(ApiKeys apiKey, short apiVersion) {
+        // Find coordinator uses identifies entities using a key type
+        if (apiKey == FIND_COORDINATOR) {
+            return true;
+        }
+        return switch (apiKey) {
+<#list messageSpecs as messageSpec>
+    <#if messageSpec.type?c_lower_case == 'request' && messageSpec.entityFields.hasAtLeastOneEntityField>
+        <#assign specName>
+            <#assign words=messageSpec.name?split("(?=[A-Z])", "r")><#list words as word>${word?c_upper_case}<#sep>_</#list>
+        </#assign>
+            case ${retrieveApiKey(messageSpec)} -> inVersion(apiVersion, ${specName?trim}_VERSIONS);
+    </#if>
+</#list>
+            default -> false;
+        };
+    }
+
+    private static boolean inVersion(short apiVersions, Set<Short> versions) {
+        return versions.contains(apiVersions);
+    }
+
+    @Override
+    public boolean shouldHandleResponse(ApiKeys apiKey, short apiVersion) {
+        return responseKeys.contains(apiKey);
+    }
+
+    public CompletionStage<RequestFilterResult> onRequest(ApiKeys apiKey,
+                                                          RequestHeaderData header,
+                                                          ApiMessage request,
+                                                          FilterContext context) {
+        var authzId = context.clientSaslContext().map(ClientSaslContext::authorizationId);
+        authzId.ifPresent(aid -> {
```

Comment:
Right now we're using the authorized id as a prefix, but in general the prefix could be any string that's defines a scope for the namespace. For example, with the `Subject` work a natural extension would be to use some configured `@Unique` `Principal ` implementation class as the source of the prefix. That's obviously a future extension, but we can rename `aid` to `prefix` or something more generic right from the start.